Prose/ProseLib/Sources/AuthenticationFeature/AuthenticationScreen.swift

```diff
@@ -15,24 +17,205 @@ public struct AuthenticationScreen: View {
     let store: Store<State, Action>
     private var actions: ViewStore<Void, Action> { ViewStore(self.store.stateless) }
 
+    @FocusState private var focusedField: State.Field?
+
     public init(store: Store<State, Action>) {
         self.store = store
     }
 
     public var body: some View {
-        WithViewStore(self.store) { viewStore in
+        WithViewStore(self.store, removeDuplicates: { $0.focusedField == $1.focusedField }) { viewStore in
+            VStack(spacing: 32) {
+                Spacer()
+                Self.header()
+                fields()
+                logInButton()
+                footer()
+                Spacer()
+            }
+            .multilineTextAlignment(.center)
+            .controlSize(.large)
+            .textFieldStyle(.roundedBorder)
+            .buttonBorderShape(.roundedRectangle)
+            .padding(.horizontal, 48)
+            .padding(.vertical, 24)
+            .synchronize(viewStore.binding(\.$focusedField), self.$focusedField)
+            .alert(self.store.scope(state: \.alert), dismiss: .alertDismissed)
+        }
+    }
+
+    private static func header() -> some View {
+        VStack(spacing: 4) {
+            Text("👋")
+                .font(.system(size: 64))
+                .padding(.bottom, 20)
+            Text("Welcome!")
+                .font(.system(size: 24, weight: .bold))
+            Text("Sign in to your chat account")
+                .foregroundColor(.secondary)
+                .font(.system(size: 16))
+        }
+    }
+
+    private func fields() -> some View {
+        WithViewStore(self.store, removeDuplicates: { $0.isLoading == $1.isLoading }) { viewStore in
```

Comment:
What is it that you want to prevent by providing a custom `removeDuplicates` handler and nesting ViewStores? It seems like an anti-pattern to me. When I remove all nested ViewStores and the `removeDuplicates` handlers it still works as expected.

Reply:
I don't like having `WithViewStore(self.store)` in forms, as it invalidates the full view rendering every time the user taps on a key. That's why I try to scope as much as possible the `viewStore`s. It's quite verbose, but TCA doesn't make it "magical" by inferring all of this, so I do it by hand.

Reply:
Can you explain why "rendering" the view on every keystroke would be a bad thing? Isn't that just normal SwiftUI behavior which is efficient in calculating diffs between state changes? How would a vanilla (read: no use of TCA) SwiftUI form look like?

Reply:
You're right, but since we can avoid re-rendering, why wouldn't we?

Note: I don't know if SwiftUI is smart enough to only re-render the views which used the properties which changed. I think it's the case, but I have no documentation article which explains it in mind.

Reply:
Should I create a very basic example to test what I have in mind?

Reply:
> You're right, but since we can avoid re-rendering, why wouldn't we?

We wouldn't because it clutters up the code and makes us do things a computer is very good at. User input as long as it isn't a continuous gesture like a drag is usually very slow. How fast can a user click or type? Would comparing a handful of values this often be a problem for a modern computer? Would we save energy? Should we spend developer time (& money) on optimizing this problem?

Reply:
> Note: I don't know if SwiftUI is smart enough to only re-render the views which used the properties which changed. I think it's the case, but I have no documentation article which explains it in mind.

Here are some:
* https://developer.apple.com/videos/play/wwdc2021/10022/
* https://www.vadimbulavin.com/swiftui-view-lifecycle/
* https://www.hackingwithswift.com/quick-start/swiftui/how-to-find-which-data-change-is-causing-a-swiftui-view-to-update
* Also interesting: All questions and answers from the WWDC 22 lounges: https://midnight-beanie-ccb.notion.site/swiftui-lounge-wwdc22-e20094b91f074398ba395c3fa245e63d (See "Assume that I keep the whole app state in the state object in my app struct?…")

So far I have limited experience with SwiftUI myself, so if you find anything let me know. But please let's not clutter up our code on assumptions. If there's a performance issue we can fix it, but let's not optimize prematurely without actually measuring.

Reply:
Thank you for the research 🙏🏻

> Would comparing a handful of values this often be a problem for a modern computer?

The thing is, as soon as a node in your tree hierarchy changes, I'm pretty sure all of the underlying nodes are rerendered (because SwiftUI cannot know if the views will take the same amount of space for example).

> User input as long as it isn't a continuous gesture like a drag is usually very slow. How fast can a user click or type?

Fast enough for developers to have to debounce key strokes, but you're right.

> But please let's not clutter up our code on assumptions. If there's a performance issue we can fix it, but let's not optimize prematurely without actually measuring.

As I told you the other day, it's because I already had the issue that I'm fixing this every time. But if you want I can stop doing it.

By the way, I can't find

> Also interesting: All questions and answers from the WWDC 22 lounges: https://midnight-beanie-ccb.notion.site/swiftui-lounge-wwdc22-e20094b91f074398ba395c3fa245e63d (See "Assume that I keep the whole app state in the state object in my app struct?…")

Reply:
> The thing is, as soon as a node in your tree hierarchy changes, I'm pretty sure all of the underlying nodes are rerendered (because SwiftUI cannot know if the views will take the same amount of space for example).

Please define "rendering". When I in a UIKit UITextField, is that "rendering" as well? Also are you talking about the actual visible rendering or about the view structure (the SwiftUI `View` struct) which is the underlying data for a rendered view. Why would an actual (drawn) view (think CALayer) be rerendered when its underlying data (think SwiftUI View) doesn't change?

> Fast enough for developers to have to debounce key strokes, but you're right.

That is out of context. Debounce to _do what_? Fire of a network request or write to the file system? Sure, debounce because it's expensive. Relayout your views? No problem, do it in the next runloop cycle (which is essentially a debounce). Compare two tiny structs?

> As I told you the other day, it's because I already had the issue that I'm fixing this every time. But if you want I can stop doing it.

That was another setup/project, wasn't it? Otherwise, please let me have a look to understand the issue.

> > Also interesting: All questions and answers from the WWDC 22 lounges: https://midnight-beanie-ccb.notion.site/swiftui-lounge-wwdc22-e20094b91f074398ba395c3fa245e63d (See "Assume that I keep the whole app state in the state object in my app struct?…")

It's filed under 'best practices'…

<img width="701" alt="Screenshot 2022-06-13 at 10 41 32" src="https://user-images.githubusercontent.com/39174/173314664-a2dddace-511b-4f56-90b2-a78a5f65c55e.png">

Reply:
> Please define "rendering". When I in a UIKit UITextField, is that "rendering" as well? Also are you talking about the actual visible rendering or about the view structure (the SwiftUI View struct) which is the underlying data for a rendered view.

I was talking about visible rendering.

> Why would an actual (drawn) view (think CALayer) be rerendered when its underlying data (think SwiftUI View) doesn't change?

Let's say you have:

```swift
HStack {
  if isDisplayed {
    Color.red
  }
  Text("A very long paragraph")
}
```

When `isDisplayed` is `false`, `Text("A very long paragraph")` is rendered with some available space, depending on its layout priority, and the `intrinsicContentSize`s of the other views.

When `isDisplayed` switches to `true`, the underlying data of `Text` is still exactly the same, SwiftUI has already rendered it, probably cached it, but it **needs** to rerender it (otherwise I'd like to understand this magic).

That's why I said

> as soon as a node in your tree hierarchy changes, I'm pretty sure all of the underlying nodes are rerendered

Do you agree?

> That was another setup/project, wasn't it? Otherwise, please let me have a look to understand the issue.

Yes, it was another project, and I had a map next to a text field.

> It's filed under 'best practices'…

Thank you, I couldn't find it with a search (even in the HTML content).

Reply:
> Let's say you have:
> 
> ```swift
> HStack {
>   if isDisplayed {
>     Color.red
>   }
>   Text("A very long paragraph")
> }
> ```
> 
> When `isDisplayed` is `false`, `Text("A very long paragraph")` is rendered with some available space, depending on its layout priority, and the `intrinsicContentSize`s of the other views.
> 
> When `isDisplayed` switches to `true`, the underlying data of `Text` is still exactly the same, SwiftUI has already rendered it, probably cached it, but it **needs** to rerender it (otherwise I'd like to understand this magic).

I don't think that's an appropriate example for what you were doing in the AuthenticationScreen. How about this…

```swift
import SwiftUI

class ViewModel: ObservableObject {
  @Published var outer = "Outer"
  @Published var inner = "Inner"
}

struct ContentView: View {
  @ObservedObject var viewModel = ViewModel()
  
  var body: some View {
    VStack {
      OuterView(text: self.viewModel.outer) {
        InnerView(text: self.viewModel.inner)
      }
      Button("Change outer") {
        self.viewModel.outer += "."
      }
      Button("Change inner") {
        self.viewModel.inner += "."
      }
    }
  }
}

struct OuterView<T: View>: View {
  let text: String
  let content: T
  
  init(text: String, @ViewBuilder content: () -> T) {
    self.text = text
    self.content = content()
  }
  
  var body: some View {
    VStack {
      let _ = Self._printChanges()
      AppKitLabel(text: self.text)
      self.content
    }
  }
}

struct InnerView: View {
  let text: String
  
  var body: some View {
    let _ = Self._printChanges()
    AppKitLabel(text: self.text)
  }
}

struct AppKitLabel: NSViewRepresentable {
  let text: String

  func makeNSView(context: Context) -> NSTextField {
    print("Init", self.text)
    return NSTextField(string: self.text)
  }
  
  func updateNSView(_ nsView: NSTextField, context: Context) {
    print("Update", self.text)
    nsView.stringValue = self.text
  }
}
```

Reply:
If you change the `InnerView` and `OuterView` as well so that this resembles more what `WithViewStore` does like this:

```swift
struct ContentView: View {
  @ObservedObject var viewModel = ViewModel()
  
  var body: some View {
    VStack {
      OuterView(viewModel: self.viewModel) {
        InnerView(viewModel: self.viewModel)
      }
      Button("Change outer") {
        self.viewModel.outer += "."
      }
      Button("Change inner") {
        self.viewModel.inner += "."
      }
    }
  }
}

struct OuterView<T: View>: View {
  @ObservedObject var viewModel: ViewModel
  let content: T
  
  init(viewModel: ViewModel, @ViewBuilder content: () -> T) {
    self.viewModel = viewModel
    self.content = content()
  }
  
  var body: some View {
    VStack {
      let _ = Self._printChanges()
      AppKitLabel(text: self.viewModel.outer)
      self.content
    }
  }
}

struct InnerView: View {
  @ObservedObject var viewModel: ViewModel
  
  var body: some View {
    let _ = Self._printChanges()
    AppKitLabel(text: self.viewModel.inner)
  }
}
```

The output looks like this…

Start:

```
OuterView: @self, @identity, _viewModel changed.
Init Outer
Update Outer
InnerView: @self, @identity, _viewModel changed.
Init Inner
Update Inner
```


Tap "Change outer":

```
OuterView: _viewModel changed.
Update Outer.
InnerView: _viewModel changed.
```

Tap "Change inner":

```
OuterView: _viewModel changed.
InnerView: _viewModel changed.
Update Inner..
```

More diffing but same amount of updates. This is why we need to have a properly sized `ViewState` for ViewStores in views that have (many) children, like [here](https://github.com/pointfreeco/swift-composable-architecture/blob/87f388bedbf1ce484e639ace9d296016a7f20812/Examples/TicTacToe/tic-tac-toe/Sources/TwoFactorUIKit/TwoFactorViewController.swift#L33).

Reply:
> I'm just wondering now, what happens if you have two `InnerView`s next to the other and you change one of them?

There's only one way to find out…